node/src/components/network/conman.rs

```diff
@@ -0,0 +1,1054 @@
+
+                    // Judge how long the connection was active.
+                    let delay = if duration > ctx.cfg.flaky_connection_threshold {
+                        rate_limited!(LOST_CONNECTION, |dropped| info!(
+                            dropped,
+                            "lost connection, will reconnect"
+                        ));
+                        ctx.cfg.successful_reconnect_delay
+                    } else {
+                        rate_limited!(LOST_FLAKY_CONNECTION, |dropped| warn!(
+                            dropped,
+                            "lost connection, but its flaky, will reconnect later"
+                        ));
+                        ctx.cfg.significant_error_backoff
+                    };
+
+                    tokio::time::sleep(delay).await;
+
+                    // After this, the loop will repeat, triggering a reconnect.
+                }
+                Err(OutgoingError::EncounteredBannedPeer(until)) => {
+                    // We will not keep attempting to connect to banned peers, put them on the
+                    // do-not-call list.
+                    break until;
+                }
+                Err(OutgoingError::FailedToCompleteHandshake(err)) => {
+                    debug!(%err, "failed to complete handshake");
+                    break Instant::now() + ctx.cfg.significant_error_backoff;
+                }
+                Err(OutgoingError::LoopbackEncountered) => {
+                    info!("found loopback");
+                    break Instant::now() + ctx.cfg.permanent_error_backoff;
+                }
+                Err(OutgoingError::ReconnectionAttemptsExhausted(err)) => {
+                    // We could not connect to the address, so we are going to forget it.
+                    rate_limited!(
+                        RECONNECTION_ATTEMPTS_EXHAUSTED,
+                        |dropped| info!(last_error=%err, dropped, "forgetting address after exhausting reconnection attempts")
+                    );
+                    return;
+                }
+                Err(OutgoingError::RpcServerError(err)) => {
+                    rate_limited!(
+                        RPC_ERROR_ON_OUTGOING,
+                        |dropped| warn!(%err, dropped, "encountered juliet RPC error")
+                    );
+                    // TODO: If there was a user error, try to extract a reconnection hint.
+                    break Instant::now() + ctx.cfg.significant_error_backoff;
+                }
+                Err(OutgoingError::ShouldBeIncoming) => {
+                    // This is "our bad", but the peer has been informed of our address now.
+                    // TODO: When an incoming connection is made (from the peer), consider clearing
+                    //       this faster.
+                    debug!("should be incoming connection");
+                    break Instant::now() + ctx.cfg.permanent_error_backoff;
```

Comment:
Shouldn't we return here instead? Any point in trying to reconnect if we shouldn't have in the first place?

Reply:
This is not reconnecting, but adding it to the do-not-call list, then exiting! The `break` value is just used to update the list, after which `fn run` ends.